File: c/knapsack.c
#include "knapsack.h"
#include <stdlib.h>
#include <string.h>

#define max(x, y) (((x) > (y)) ? (x) : (y))

unsigned int maximum_value(unsigned int maximum_weight, const item_t *items,
                           size_t item_count) {
  unsigned int *dp = malloc(sizeof(unsigned int) * (maximum_weight + 1));
  memset(dp, 0, sizeof(unsigned int) * (maximum_weight + 1));

  for (size_t i = 0; i < item_count; i++) {
    for (unsigned int j = maximum_weight; j >= items[i].weight; j--) {
      dp[j] = max(dp[j], dp[j - items[i].weight] + items[i].value);
    }
  }

  unsigned int ans = dp[maximum_weight];

  free(dp);

  return ans;
}

File: elixir/knapsack.ex
defmodule Knapsack do
  @doc """
  Return the maximum value that a knapsack can carry.
  """
  @spec maximum_value(items :: [%{value: integer, weight: integer}], maximum_weight :: integer) ::
          integer
  def maximum_value(items, maximum_weight) do
    items
    |> Enum.reduce(%{}, fn %{value: value, weight: weight}, acc ->
      maximum_weight..weight//-1
      |> Enum.reduce(acc, fn w, acc ->
        v = max((acc[w - weight] || 0) + value, acc[w] || 0)
        Map.put(acc, w, v)
      end)
    end)
    |> Map.get(maximum_weight, 0)
  end
end
